.github/CONTRIBUTING.md

```diff
@@ -17,40 +16,103 @@ We prefer questions and support requests be posted in the [forum](http://forum.t
 
 __Please provide as much information as possible to help us solve problems and answer questions better!__
 
-Commits and Commit Messages
----------------------------
-Follow these guidelines when creating public commits and writing commit messages.
+## PR Title / Commit Message Guidelines
+
+We try to follow the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0-beta.2/) style for commit messages and pull request titles.
+This leads to **more readable messages** that are easy to follow when looking through the **project history**.
+But also, we use the git commit messages to **generate the Terasology change log**.
+
+### Commit Message Format
+Each commit message consists of a **header**, a **body** and a **footer**.  The header has a special
+format that includes a **type**, a **scope** and a **subject**:
+
+```
+<type>(<scope>): <subject>
+<BLANK LINE>
+<body>
+<BLANK LINE>
+<footer>
+```
+
+The **header** is mandatory and the **scope** of the header is optional.
+
+Any line of the commit message cannot be longer 100 characters! This allows the message to be easier
+to read on GitHub as well as in various git tools.
+
+The footer should contain a [closing reference to an issue](https://help.github.com/articles/closing-issues-via-commit-messages/) if any.
+
+Samples: (even more [samples](https://github.com/angular/angular/commits/master))
+
+```
+docs(changelog): update changelog to beta.5
+```
+```
+fix(release): need to depend on latest rxjs and zone.js
+
+The version in our package.json gets copied to the one we publish, and users need the latest of these.
+```
+
+### Revert
+If the commit reverts a previous commit, it should begin with `revert: `, followed by the header of the reverted commit. In the body it should say: `This reverts commit <hash>.`, where the hash is the SHA of the commit being reverted.
+
+### Type
+Must be one of the following:
+
+* **build**: Changes that affect the build system or external dependencies (example scopes: gradle, git, idea)
+* **ci**: Changes to our CI configuration files and scripts (example scopes: Jenkins)
+* **docs**: Documentation only changes
+* **feat**: A new feature
+* **fix**: A bug fix for the game
+* **perf**: A code change that improves performance of the game
+* **refactor**: A code change that neither fixes a bug nor adds a feature
+* **style**: Changes that do not affect the meaning of the code (white-space, formatting, copyright, etc)
+* **test**: Adding missing tests or correcting existing tests
+
+### Scope
+The scope should be the name of the package(s) affected (as perceived by the person reading the changelog generated from commit messages).
```

Comment:
This is one I've gone back and forth on over time. There certainly use cases it serves well, and then there are other things like `git annotate` where it would just be showing you the name of the package you already know you're looking at.

and that information _can_ be automatically derived from the filenames present in the commit.

Reply:
yeah, I'm not sure here either... I'd like to see this as an optional part we can fill in if it makes sense. For instance, for migrating JOML, it might be something like
```
refactor(world.block): migrate FooEvent and BarComponent
```
which narrows down the affected package without the need to look up the content of the respective commit/PR. It also can help to narrow down the semantic change in a PR from the follow-up changes necessary to adjust...